### DISCUSSION TITLE
SPA with client routing, passToClient error

### DISCUSSION THREAD

**Discussion by jscottsf** [vikejs/vike#1041] [Discussion]
Trying to get an SPA with VPS client routing going. Have looked over all of the examples.

When it runs and I change route (client-side navigation), a browser error is thrown insisting that I add `_routeMatches`, `_routingProvidedByOnBeforeRouteHook`, and `routeParams` to my `passToClient`. So this ends up working.

```
export default {
  clientRouting: true,
  meta: {
    Layout: {
      env: 'client-only'
    },
    Page: {
      env: 'client-only'
    }
  },
  passToClient: [
    '_routeMatches',
    '_routingProvidedByOnBeforeRouteHook',
    'routeParams',
    'documentProps',
    'pageProps'
  ],
  prerender: false,
  prefetchStaticAssets: 'viewport'
}
```

It's very odd.

Also, I noticed that the generated `vite-plugin-ssr.json` in client output indicates that client routing is `false`. Is this correct?

```
{
  "version": "0.4.133",
  "usesClientRouter": false,
  "manifestKeyMap": {},
  "baseServer": "/",
  "baseAssets": "/",
  "includeAssetsImportedByServer": true
}
```

**Comment by jscottsf** [comment]
For more context, I am serving the app using `serve` or `http-server` to simulate running on a CDN as static.

**Comment by brillout** [comment]
> a browser error is thrown insisting that I add `_routeMatches`, `_routingProvidedByOnBeforeRouteHook`, and `routeParams` to my `passToClient`
 

> `vite-plugin-ssr.json` in client output indicates that client routing is `false`


Something seems off with your setup. Can you publish a reproduction? I'll have a look at it.

**Reply by jscottsf** [reply]
Think I found the issue. I will commit a repo, or can send a zip if you want.

**Reply by jscottsf** [reply]
Here is the repo https://github.com/jscottsf/vps-spa-jscottsf

I stripped down our company boilerplate to the basics.

See the NOTE in `renderer/+config.h.js`.

It seems to work with onBeforeRender set to `client-only`, however, `"usesClientRouter": false,` remains either way. If there is any significance to that.

Thank you!

P.S. Have some additional questions later re: hosting statically and using parameterized routes without prerendering.

**Reply by brillout** [reply]
> `"usesClientRouter": false,` remains

That's expected because the V1 design [doesn't use it anymore](https://github.com/brillout/vite-plugin-ssr/commit/33e6c307615e59bf5c141768c947e9725bb632d8).

> a browser error is thrown insisting that I add `_routeMatches`, `_routingProvidedByOnBeforeRouteHook`, and `routeParams` to my `passToClient`

I can confirm, it seems to be an issue with how vite-plugin-ssr handles Vue's reactivity. I'll further dig into it later.

> onBeforeRender set to `client-only`

It's actually a new possibility enabled by the V1 design. I never tested it so you may encounter issues.

> However, the onBeforeRender function is not called client-side on
> initial page load. It's called only on subsequent client route
> changes. This make it difficult to do initial page fetching in
> onBeforeRender for an SPA on static hosting.

Makes sense, I created a ticket for it https://github.com/brillout/vite-plugin-ssr/issues/1043.

> additional questions later re: hosting statically and using parameterized routes without prerendering

👍

Btw: would your company be up for [sponsoring](https://github.com/sponsors/brillout)?

**Reply by jscottsf** [reply]
Thank you for looking into this!

Already an individual sponsor with my own company, but only base level. Currently using VPS for one small contract and was evaluating use for something with UC Berkeley. One team concern with UCB was not enough maintainers for the project -- yet this is far easier to use than dealing the hellscape of Nuxt. Having a lightweight alternative is needed and appreciated.

**Reply by brillout** [reply]
> I can confirm, it seems to be an issue with how vite-plugin-ssr handles Vue's reactivity. I'll further dig into it later.

Fixed.

> I created a ticket for it #1043.

Done. (Although I didn't test it: let me know if it doesn't work.)

Both released in `vite-plugin-ssr@0.4.134-commit-ddf2e39`.

> concern with UCB was not enough maintainers for the project

That's actually an argument for sponsoring: once I get enough sponsors to cover my livings costs (almost arleady the case), then there won't be much reasons for me to stop, as developing vite-plugin-ssr has been the most thrilling endeavour in my career.

**Reply by jscottsf** [reply]
Thx for looking into this!

Tested `onBeforeRender.env`and works perfectly for `client-only` and `server-only`. For `server-and-client` it only fires `onBeforeRender` client-side on the subsequent route changes, not on the initial visit.

Tested against `0.4.135`.

**Reply by jscottsf** [reply]
The `Error: [vite-plugin-ssr][Wrong Usage] pageContext.` errors are no longer being thrown. Thank you!

**Reply by jscottsf** [reply]
Since this will be used in a to-be production site soon, and partially in a new project. I'll be bumping my personal contribution to $20/mo for a few months. Still on a budget here. Have you explored channels to gain greater adoption (and hence funding) -- or mainly organic?

**Reply by brillout** [reply]
> For `server-and-client` it only fires `onBeforeRender` client-side on the subsequent route changes, not on the initial visit.

That's on purpose: on the initial visit, it's called only on the server-side. It's always called and the only question is where.

> I'll be bumping my personal contribution to $20/mo for a few months

Thanks!

> Have you explored channels to gain greater adoption (and hence funding) -- or mainly organic?

Mainly organic so far. I'm open for ideas for more channels though.
